src/hosting/implicitInit.js

```diff
@@ -1,17 +1,41 @@
 "use strict";
 
 var fs = require("fs");
-
-var { fetchWebSetup } = require("../fetchWebSetup");
+var { fetchWebSetup, getCachedWebSetup } = require("../fetchWebSetup");
+var utils = require("../utils");
+var logger = require("../logger");
 
 var INIT_TEMPLATE = fs.readFileSync(__dirname + "/../../templates/hosting/init.js", "utf8");
 
-module.exports = function(options) {
-  return fetchWebSetup(options).then(function(config) {
-    var configJson = JSON.stringify(config, null, 2);
-    return {
-      js: INIT_TEMPLATE.replace("{/*--CONFIG--*/}", configJson),
-      json: configJson,
-    };
-  });
+module.exports = async function(options) {
+  let config;
+  try {
+    config = await fetchWebSetup(options);
+  } catch (e) {
+    logger.debug("fetchWebSetup error: " + e);
+  }
+
+  if (!config) {
+    config = getCachedWebSetup(options);
+    if (config) {
+      utils.logLabeledWarning(
+        "hosting",
+        "You are offline, using web app configuration from cache."
+      );
+    }
+  }
+
+  if (!config) {
+    config = undefined;
+    utils.logLabeledWarning(
+      "hosting",
+      "You are offline and there is no cached configuration on this machine. You must call firebase.initializeApp({...}) in your code before using Firebase."
```

Comment:
I **think** a user would also hit this warning when they either have never logged in, or have logged in and don't have permissions.

In that case it is confusing to state the user is offline. Perhaps instead of "You are offline" it would be clearer to say "Cannot access online project configuration" in both of these messages?

Reply:
Good catch!  Let me clarify.